syntax = "proto3";

package denden;

option go_package = "github.com/strawpot/denden/cli/gen/denden";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ---------------------------------------------------------------------------
// Service
// ---------------------------------------------------------------------------

service Denden {
  // Synchronous request dispatch (ask_user or delegate).
  rpc Send (DenDenRequest) returns (DenDenResponse);

  // Health check.
  rpc Status (StatusRequest) returns (StatusResponse);
}

// ---------------------------------------------------------------------------
// Request envelope
// ---------------------------------------------------------------------------

message DenDenRequest {
  string denden_version = 1;
  string request_id = 2;
  Trace trace = 3;

  oneof payload {
    AskUserPayload ask_user = 10;
    DelegatePayload delegate = 11;
  }
}

message Trace {
  string run_id = 1;
  string agent_instance_id = 2;
  string parent_agent_instance_id = 3;
  google.protobuf.Timestamp created_at = 4;
}

// ---------------------------------------------------------------------------
// ask_user (synchronous)
// ---------------------------------------------------------------------------

message AskUserPayload {
  string question = 1;
  repeated string choices = 2;
  string default_value = 3;
  string why = 4;
  Format response_format = 5;
}

// ---------------------------------------------------------------------------
// delegate (synchronous; orchestrator executes internally)
// ---------------------------------------------------------------------------

message DelegatePayload {
  string delegate_to = 1;
  Task task = 2;
}

message Task {
  string text = 1;
  repeated string artifact_refs = 2;
  google.protobuf.Struct extra = 3;
  Format return_format = 4;
}

enum Format {
  TEXT = 0;
  JSON = 1;
}

// ---------------------------------------------------------------------------
// Response envelope
// ---------------------------------------------------------------------------

message DenDenResponse {
  string denden_version = 1;
  string request_id = 2;
  ResponseStatus status = 3;
  ErrorDetail error = 4;

  oneof result {
    AskUserResult ask_user_result = 10;
    DelegateResult delegate_result = 11;
  }
}

enum ResponseStatus {
  OK = 0;
  DENIED = 1;
  ERROR = 2;
}

message ErrorDetail {
  string code = 1;
  string message = 2;
  bool retryable = 3;
}

// ---------------------------------------------------------------------------
// ask_user result
// ---------------------------------------------------------------------------

message AskUserResult {
  oneof content {
    string text = 1;
    google.protobuf.Struct json = 2;
  }
}

// ---------------------------------------------------------------------------
// delegate result
// ---------------------------------------------------------------------------

message DelegateResult {
  Format output_format = 1;
  google.protobuf.Struct output = 2;
  string summary = 3;
}

// ---------------------------------------------------------------------------
// Status
// ---------------------------------------------------------------------------

message StatusRequest {}

message StatusResponse {
  int64 uptime_seconds = 1;
  int32 active_agents = 2;
}
